### PR TITLE
Implement kmeans.py and experiment kmeans++ in test

### PR DESCRIPTION
**Pull request by MrWickham**
After some experiment, I found out that the test will have a chance to past with kmeans++ after setting the range in for-loop in k=2 and k= 3 to 800 and 600. Still, it seems like the result isn't deterministic as the test will still fail from time to time.

### CODE REVIEW

**Comment by gallettilance** [comment]
@MrWickham if you can also squash your commits that would be great

**Comment by MrWickham** [comment]
done

File: 02-library/tests/test_kmeans_pp.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    assert clustered == [expected_clustering]
+
+
+@pytest.mark.parametrize('datasetPath,expected1,expected2', [
+    ("tests/test_files/dataset_1.csv",
+     "tests/test_files/dataset_1_k_is_2_0.csv",
+     "tests/test_files/dataset_1_k_is_2_1.csv"),
+])
+def test_kmeans_when_k_is_2(datasetPath, expected1, expected2):
+    # random.seed(1)
```
Comment (gallettilance) [inline 02-library/tests/test_kmeans_pp.py]: You can remove this line if it's not needed to pass the tests

File: 02-library/tests/test_kmeans_pp.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+
+@pytest.mark.parametrize('datasetPath,expected1,expected2,expected3', [
+    ("tests/test_files/dataset_1.csv",
+     "tests/test_files/dataset_1_k_is_3_0.csv",
+     "tests/test_files/dataset_1_k_is_3_1.csv",
+     "tests/test_files/dataset_1_k_is_3_2.csv"),
+])
+def test_kmeans_when_k_is_3(datasetPath, expected1, expected2, expected3):
+    # random.seed(1)
```
Comment (gallettilance) [inline 02-library/tests/test_kmeans_pp.py]: same here

File: 02-library/cs506/kmeans.py [hunk 3]
```diff
@@ -83,4 +115,4 @@ def k_means(dataset, k):
 
 def k_means_pp(dataset, k):
     k_points = generate_k_pp(dataset, k)
-    return _do_lloyds_algo(dataset, k_points)
+    return _do_lloyds_algo(dataset, k_points)
```
Comment (gallettilance) [inline 02-library/cs506/kmeans.py]: Nit: please add a new line at the end of the file

File: 02-library/cs506/kmeans.py [hunk 4]
```diff
@@ -83,4 +115,4 @@ def k_means(dataset, k):
 
 def k_means_pp(dataset, k):
```
Comment (gallettilance) [inline 02-library/cs506/kmeans.py]: Can you add error handling for when `k` is not in
```python
range(1,len(dataset))
```
Applies for k_means too

File: 02-library/cs506/kmeans.py [hunk 5]
```diff
@@ -44,23 +55,44 @@ def distance(a, b):
     """
     Returns the Euclidean distance between a and b
     """
-    raise NotImplementedError()
+    return sum([(a - b) ** 2 for a, b in zip(a, b)]) ** 0.5
+
+def distance_squared(a, b):
+    return sum([(a - b) ** 2 for a, b in zip(a, b)])
```
Comment (gallettilance) [inline 02-library/cs506/kmeans.py]:
```suggestion
    return distance(a, b) ** 2
```

File: 02-library/cs506/kmeans.py [hunk 6]
```diff
@@ -21,8 +24,16 @@ def update_centers(dataset, assignments):
     Compute the center for each of the assigned groups.
     Return `k` centers in a list
     """
-    raise NotImplementedError()
+    k = max(assignments) + 1
+    clusters = [[] for i in range(k)]
+    for i, index in enumerate(assignments):
```
Comment (gallettilance) [inline 02-library/cs506/kmeans.py]: it's not very clear what the distinction is between `i` and `index`. Can you make the names more descriptive? For example:
```suggestion
    for pointIndex, pointAssignment in enumerate(assignments):
```

**Comment by MrWickham** [comment]
While minor issues have been fixed, in cli.py I initially intend to have codes like below:
```python
    def kmeans_template(self, args, func):
        """Run the kmeans command
        """
        import csv
        dataset = read.read_csv(args.dataset_file)
        clustering = func(dataset=dataset, k=int(args.k))
        cost = kmeans.cost_function(clustering)

        for _ in range(100):
            new_clustering = func(dataset=dataset, k=int(args.k))
            new_cost = kmeans.cost_function(clustering)
            if new_cost < cost:
                clustering = new_clustering
                cost = new_cost

        for assignment in clustering.keys():
            file_name = str(args.dataset).split(".")[0]+"_k_is_"+args.k+"_"+str(assignment)+".csv"
            with open(file_name, "w") as f:
                writer = csv.writer(f)
                print("assignement ", assignment, " is: ", clustering[assignment])
                writer.writerows(clustering[assignment])
            f.close()

    def kmeans(self, args):
        kmeans_template(self, args, kmeans.k_means)

    def kmeans_pp(self, args):
        kmeans_template(self, args, kmeans.k_means_pp)
```

But that didn't work. So I changed kmeans() and kmeans_pp() to this:

```python
    def kmeans(self, args, func=kmeans_template):
        func(self, args, kmeans.k_means)

    def kmeans_pp(self, args, func=kmeans_template):
        func(self, args, kmeans.k_means_pp)
``` 

Is that ok? Or is there a better way to achieve this?

**Comment by gallettilance** [comment]
I just re-ran the tests but please force push when they fail to re-run the tests until they pass :)

**Comment by gallettilance** [comment]
@MrWickham please get the tests to pass. Feel free to change the csv for the kmeans_pp.

**Comment by MrWickham** [comment]
I tried to debug today, then I found out there might be a issue with the order of points in each cluster, I'll try to see if I can fix this with set() or sort().

**Pull request merged**